File: lib/dog.dart
@JS()
library dog;

import 'package:js/js.dart';

@JS()
class Dog {
  external factory Dog(String name, int age);
  external String get name;
  external int get age;
  external void bark();
  external void jump(Function(int height) func);
  external void sleep(Options options);
}

@JS()
@anonymous
class Options {
  external bool get bed;
  external String get hardness;
  external factory Options({bool bed, String hardness});
}

File: web/Dog.js
class Dog {
    constructor(name, age) {
        this._name = name;
        this._age = age;
        console.log("Dog created")
    }

    get name() {
        return this._name;
    }

    get age() {
        return this._age;
    }

    bark() {
        console.log(`${this._name}:${this._age}:: Woof!`);
    }

    sleep(options) {
        if (options.bed) {
            console.log(`${this._name} is sleeping on a ${options.hardness} bed.`);
        } else {
            console.log(`${this._name} is sleeping on the floor. :(`);
        }
    }
}


window.Dog = Dog;
